query.go

```diff
@@ -111,145 +121,125 @@ loop:
-			})
-
-		// The lookup terminates when the initiator has queried and gotten responses from the k
-		// closest nodes it has heard about.
-		if len(peersToQuery) == 0 {
+		select {
+		case sawPeers := <-ch:
+			q.populatePeers(sawPeers)
+		case <-pathCtx.Done():
+			q.terminate()
 			return
 		}
 
-		// Of the k nodes the initiator has heard of closest to the target,
-		// it picks α that it has not yet queried and resends the FIND NODE RPC to them.
-		numQuery := q.dht.alpha
-
-		// However, If a round of RPC's fails to return a node any closer than the closest already heard about,
-		// the initiator resends the RPC'S to all of the k closest nodes it has
-		// not already queried.
-		if !foundCloser {
-			numQuery = len(peersToQuery)
-		} else if pqLen := len(peersToQuery); pqLen < numQuery {
-			// if we don't have α peers, pick whatever number we have.
-			numQuery = pqLen
+		// termination is triggered on end-of-lookup conditions or starvation of unused peers
+		if q.readyToTerminate() {
+			q.terminate()
+			return
 		}
 
-		// reset foundCloser to false for the next round of RPC's
-		foundCloser = false
-
-		queryResCh := make(chan *queryResult, numQuery)
-		resultsReceived := 0
+		// if all "threads" are busy, wait until someone finishes
+		if q.queryPeers.NumWaiting() >= alpha {
+			continue
+		}
 
-		// send RPC's to all the chosen peers concurrently
-		for _, p := range peersToQuery[:numQuery] {
-			go func(p peer.ID) {
-				queryResCh <- q.queryPeer(p)
-			}(p)
+		// spawn new queries, up to the parallelism allowance
+		for j := 0; j < alpha-q.queryPeers.NumWaiting(); j++ {
+			q.spawnQuery(ch)
 		}
+	}
+}
 
-	loop:
-		// wait for all outstanding RPC's to complete before we start the next round.
-		for {
-			select {
-			case res := <-queryResCh:
-				foundCloser = foundCloser || res.foundCloserPeer
-				resultsReceived++
-				if resultsReceived == numQuery {
-					break loop
-				}
-			case <-q.ctx.Done():
-				return
-			}
+// spawnQuery starts one query, if an available seen peer is found
+func (q *query) spawnQuery(ch chan<- *queryUpdate) {
+	if peers := q.queryPeers.GetSortedSeen(); len(peers) == 0 {
+		return
+	} else {
+		q.queryPeers.MarkWaiting(peers[0])
+		go q.queryPeer(ch, peers[0])
+	}
+}
+
+func (q *query) readyToTerminate() bool {
+	// if termination has already been determined, the query is considered terminated forever,
+	// regardless of any change to queryPeers that might occur after the initial termination.
+	if q.terminated {
+		return true
+	}
+	// give the application logic a chance to terminate
+	if q.stopFn() {
+		return true
+	}
+	if q.isStarvationTermination() {
+		return true
+	}
+	if q.isLookupTermination() {
+		return true
+	}
+	return false
+}
+
+// From the set of all nodes that are not unreachable,
+// if the closest beta nodes are all queried, the lookup can terminate.
+func (q *query) isLookupTermination() bool {
+	var peers []peer.ID
+	peers = q.queryPeers.GetClosestNotUnreachable(q.dht.beta)
```

Comment:
This will also return true if `GetClosestNotUnreachable ` does NOT return any peers here which I think is the starvation condition. Do you want check for len(peers) > 0 to make this explicit ?

Reply:
This is as intended.

Reply:
See comment earlier. Having the starvation condition separately, allows me to not worry about it here. Which in turn makes the code here simpler.